File: elixir/meetup/lib/meetup.ex
defmodule Meetup do
  @moduledoc """
  Calculate meetup dates.
  """

  @type weekday ::
          :monday
          | :tuesday
          | :wednesday
          | :thursday
          | :friday
          | :saturday
          | :sunday

  @type schedule :: :first | :second | :third | :fourth | :last | :teenth

  @doc """
  Calculate a meetup date.

  The schedule is in which week (1..4, last or "teenth") the meetup date should
  fall.
  """
  @spec meetup(pos_integer, pos_integer, weekday, schedule) :: :calendar.date()
  def meetup(year, month, weekday, schedule) do
    week_index = get_week_index(weekday)
    day = get_date_range(schedule)
    |> Enum.find(&get_date(week_index, year, month, &1))
    Date.new(year, month, day)
    |> elem(1)
  end

  defp get_week_index(:monday), do: 1
  defp get_week_index(:tuesday), do: 2
  defp get_week_index(:wednesday), do: 3
  defp get_week_index(:thursday), do: 4
  defp get_week_index(:friday), do: 5
  defp get_week_index(:saturday), do: 6
  defp get_week_index(:sunday), do: 7

  defp get_date_range(:first), do: 1..7
  defp get_date_range(:second), do: 8..14
  defp get_date_range(:third), do: 15..21
  defp get_date_range(:fourth), do: 22..28
  defp get_date_range(:last), do: 31..22
  defp get_date_range(:teenth), do: 13..19

  defp get_date(index, year, month, day) do
    case Date.new(year, month, day) do
      {:ok, date} -> Date.day_of_week(date) === index
      _ -> false
    end
  end

end

File: c/meetup/meetup.c
#include "meetup.h"
#include <string.h>
#include <time.h>

static int day_to_tm_wday(const char *day) {
  const char *days[] = {"Sunday",   "Monday", "Tuesday", "Wednesday",
                        "Thursday", "Friday", "Saturday"};
  for (size_t i = 0; i < 7; ++i) {
    if (strcmp(day, days[i]) == 0) {
      return i;
    }
  }
  return -1;
}

static int week_offset_in_tm_mday(const char *week) {
  const char *offsets[] = {"first", "second", "third", "fourth"};
  for (size_t i = 0; i < 5; ++i) {
    if (strcmp(week, offsets[i]) == 0) {
      return i * 7;
    }
  }
  return -1;
}

int meetup_day_of_month(unsigned int year, unsigned int month, const char *week,
                        const char *day_of_week) {
  struct tm time = {
      .tm_year = year - 1900,
      .tm_mon = month - 1,
      .tm_mday = 1,
  };
  mktime(&time);
  int target_day = day_to_tm_wday(day_of_week);
  if (strcmp(week, "teenth") == 0) {
    while (time.tm_mday < 13) {
      ++time.tm_mday;
      time.tm_wday = (time.tm_wday + 1) % 7;
    }
  } else if (strcmp(week, "last") == 0) {
    time = (struct tm){
        .tm_year = year - 1900,
        .tm_mon = month,
        .tm_mday = 0,
    };
    mktime(&time);
    while (time.tm_wday != target_day) {
      --time.tm_mday;
      time.tm_wday = time.tm_wday - 1 < 0 ? 6 : time.tm_wday - 1;
    }
    return time.tm_mday;
  } else {
    int offset = week_offset_in_tm_mday(week);
    for (int i = 0; i < offset; ++i) {
      ++time.tm_mday;
      time.tm_wday = (time.tm_wday + 1) % 7;
    }
  }
  while (time.tm_wday != target_day) {
    ++time.tm_mday;
    time.tm_wday = (time.tm_wday + 1) % 7;
  }
  return time.tm_mday;
}
